daffodil-runtime2/src/main/resources/c/libruntime/parsers.c

```diff
@@ -142,36 +136,38 @@ define_parse_endian_integer(le, uint, 32)
 define_parse_endian_integer(le, uint, 64)
 define_parse_endian_integer(le, uint, 8)
 
-// Define function to parse fill bytes until end position is reached
+// Parse fill bytes until end position is reached
 
 void
 parse_fill_bytes(size_t end_position, PState *pstate)
 {
-    while (!pstate->error_msg && pstate->position < end_position)
+    union
     {
-        char   buffer;
-        size_t count = fread(&buffer, 1, sizeof(buffer), pstate->stream);
+        char c_val[1];
+    } buffer;
 
-        pstate->position += count;
-        if (count < sizeof(buffer))
-        {
-            pstate->error_msg = eof_or_error_msg(pstate->stream);
-        }
+    while (pstate->position < end_position)
+    {
+        read_stream_update_position;
     }
 }
 
-// Define function to validate number is same as fixed value after parse
+// Validate parsed number is same as fixed value
 
 void
 parse_validate_fixed(bool same, const char *element, PState *pstate)
 {
-    UNUSED(element); // because managing strings hard in embedded C
-    if (!pstate->error_msg && !same)
+    if (!same)
     {
-        // Error message would be easier to assemble and
-        // internationalize if we used an error struct with multiple
-        // fields instead of a const char string.
-        pstate->error_msg = "Parse: Value of element does not match value of "
-                            "its 'fixed' attribute";
+        Diagnostics *validati = need_diagnostics();
+        pstate->validati = validati;
+
+        if (validati->length <
+            sizeof(validati->array) / sizeof(*validati->array))
+        {
+            Error *error = &validati->array[validati->length++];
+            error->code = ERR_FIXED_VALUE;
+            error->s = element;
+        }
```

Comment:
This is the compile-time array length calculation I was talking about.  I'll replace this if statement with an add_diagnostic call as you suggested.  Do you really want add_diagnostic to stop the program as soon as the array fills up, or simply stop adding any more diagnostics to the array?

```c
        const Error error = {ERR_FIXED_VALUE, {element}};
        add_diagnostic(validati, &error);
```